### ISSUE TITLE
FlxGroup.as sort() crashes if it encounters another FlxGroup

### ISSUE THREAD

**Issue by FlixelCommunityBot** [FlixelCommunity/flixel#28]
Issue [#204](https://github.com/AdamAtomic/flixel/issues/204) by: [ChainedLupine](http://github.com/ChainedLupine)

Example:

```
var group1:FlxGroup  = new FlxGroup ;
var group2:FlxGroup = new FlxGroup ;

group1.add (group2) ;

group1.sort() ;  // crash
```

In FlxGroup.sort, at sortHandler, it should check to see if it's trying to sort a FlxGroup.  Something like:

```
if (Obj1 is FlxGroup || Obj2 is FlxGroup)
  return 0 ;
```

**Comment by moly** [comment]
The problem is that the default sortIndex is "y", and FlxGroup doesn't have a y property. The suggested fix won't really work because the function will still crash for other objects that don't have a y property (like FlxBasic). It would also leave the FlxGroups randomly unsorted in the parent FlxGroup.

**Comment by IQAndreas** [comment]
How about adding an error message inside of `FlxGroup::sortHandler()` along these lines:

``` as3
protected function sortHandler(Obj1:FlxBasic,Obj2:FlxBasic):int
{
    if (!(_sortIndex in Obj1) || !(_sortIndex in Obj2))
    {
        var objectName:getQualifiedClassName(Obj1); // TODO: Obj2
        FlxG.log("WARNING: Cannot sort FlxGroup. The object " + Obj1 + " does not contain the property '" + _sortIndex + "'.");
        return 0; // No particular sort order
    }

    if(Obj1[_sortIndex] < Obj2[_sortIndex])
        return _sortOrder;
    else if(Obj1[_sortIndex] > Obj2[_sortIndex])
        return -_sortOrder;
    return 0;
}
```

This will clearly alert the developer to what is wrong and why, rather than leave it up to Flash Player to throw an ugly, non-descriptive error.

Though, it may be better checking through all members first in the `sort` function rather than in `sortHandler()` (since `sortHandler` will iterate through each object several times, causing it to show up in the log more than once). Doing so will also save on performance.

The following line:

``` as3
return 0; // No particular sort order
```

Is there so the code will match Flash's way of handling "unsortable items":

> If either of the elements being compared does not contain the field that is specified in the fieldName parameter, the field is assumed to be set to undefined, and the elements are placed consecutively in the sorted array in no particular order. (via [Array::sortOn()](http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/Array.html#sortOn%28%29))

**Comment by Dovyski** [comment]
I really liked the idea @IQAndreas described. The code will continue to sort without exceptions and "non-sortable" objects will be place in no particular order (compliant with Flash's way).

As he suggested, the test should be performed in the `sort()` function rather than in `sortHandler()` to avoid lots of log entries (and increase performance).

**Comment by moly** [comment]
Is it possible just to get rid of `sortHandler` and use `members.sortOn()` instead?

**Comment by Dovyski** [comment]
That would be a good idea, but according to Flash docs related to  `sortOn()`:

> All of the objects have at least one property in common

The reason why `FlxGroup#sort()` crashes is exactly because the objects do not have one property in common. I think Flash will thrown an exception if `sortOn()` encounters an object without the sorting property.

**Comment by moly** [comment]
As Andreas quoted:

> If either of the elements being compared does not contain the field that is specified in the fieldName parameter, the field is assumed to be set to undefined, and the elements are placed consecutively in the sorted array in no particular order.

**Comment by Dovyski** [comment]
Sorry, I didn't see it was related to `sortOn`. On that light I see no reason why we should not replace `_sortHandler()` with `members.sortOn()`.

I guess Adam originally created that method to allow subclasses to override the sorting mechanism, but it can still be achieved by overriding the `sort()` method.

**Comment by IQAndreas** [comment]
I'm all for using `Array.sortOn()` instead.

> I guess Adam originally created that method to allow subclasses to override the sorting mechanism

If this is really important to developers, we could also do something like this:

``` as3
protected var sortHandler:Function = null;
// ...
public function sort(Index:String="y",Order:int=ASCENDING):void
{
    if (sortHandler != null)
    {
        _sortIndex = Index;
        _sortOrder = Order;
        members.sort(sortHandler);
    }
    else
    {
        members.sortOn(Index, <order>);
    }
}
```

**Comment by IQAndreas** [comment]
Though, on second thought, you are right, overriding `sort()` may just be easier for the developers, and will give them more control over what they want to do.

**Comment by Dovyski** [comment]
Yes :) I've just assigned this issue to myself. If any of you already have a fix, please let me know and I will reassign the issue.

**Comment by Dovyski** [comment]
I think I've found a bug related to `Array.sortOn()`. I replaced Flixel `_sortHandler()` with the built-in `Array.sortOn()` method as we discussed, but it still crashes:

```
ReferenceError: Error #1069: Property y not found on org.flixel.FlxGroup and there is no default value.
    at Array$/_sortOn()
    at Array/http://adobe.com/AS3/2006/builtin::sortOn()
    at org.flixel::FlxGroup/sort()[C:\Temp\FlixelCommunity\Versions\flixel-dovyski\org\flixel\FlxGroup.as:308]
    at PlayState/create()[C:\Temp\FlixelCommunity\Testing\src\PlayState.as:70]
    at org.flixel::FlxGame/switchState()[C:\Temp\FlixelCommunity\Versions\flixel-dovyski\org\flixel\FlxGame.as:499]
    at org.flixel::FlxGame/step()[C:\Temp\FlixelCommunity\Versions\flixel-dovyski\org\flixel\FlxGame.as:544]
    at org.flixel::FlxGame/onEnterFrame()[C:\Temp\FlixelCommunity\Versions\flixel-dovyski\org\flixel\FlxGame.as:453]

```

This result does not comply with Flash docs:

> If either of the elements being compared does not contain the field that is specified in the fieldName parameter, the field is assumed to be set to undefined, and the elements are placed consecutively in the sorted array in no particular order

Any thoughts?

**Comment by IQAndreas** [comment]
> This result does not comply with Flash docs

That's odd, I'm getting the same results. I wonder if it's a flaw in the documentation or in the Array.

Out of curiosity, I did some more testing and posted the results in a thread (was too lazy to write up an entire blog post on the matter)
http://www.kirupa.com/forum/showthread.php?374924-Array-sortOn()-does-not-work-as-the-documentation-says-it-should&p=2646165#post2646165

I guess we will have to use a custom sorting function anyway.

**Comment by Dovyski** [comment]
I saw your tests and `sortOn()` looks really odd indeed. I will continue to use the custom sorting function and handle any missing sort property according to the `sortOn()` docs (if `Array` is not compliant with it, that's Adobe's fault :)

**Comment by Dovyski** [comment]
I've just sent a pull request that fixes this issue. After some thought, I decided to make the sorting method silent (it does not produce any log when the sorting property is missing).

I thought it could generate too many log entries if the sorting property was missing. All objects with missing properties are placed at the end of the list (no particular order among them).

**Comment by Dovyski** [comment]
Fixed by commit 3a666c6d163348d01830d4921f1eb256a2de04b3
